compiler/backend_inkwell/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+                    self.builder.build_store(arr_alloc, v);
+                    let cast = self.builder.build_bitcast(
+                        arr_alloc,
+                        i8_type.ptr_type(AddressSpace::default()),
+                        "",
+                    );
+                    let make_candy_text = self.module.get_function("make_candy_text").unwrap();
+                    let call = self.builder.build_call(make_candy_text, &[cast.into()], "");
+
+                    self.builder.build_store(
+                        global.as_pointer_value(),
+                        call.try_as_basic_value().unwrap_left(),
+                    );
+
+                    global.set_initializer(&candy_type_ptr.const_null());
+                    self.values.insert(*responsible_parameter, Rc::new(global));
+
+                    let candy_type_ptr = self
+                        .module
+                        .get_struct_type("candy_type")
+                        .unwrap()
+                        .ptr_type(AddressSpace::default());
+
+                    let captured_ids: Vec<_> = expr
+                        .captured_ids()
+                        .into_iter()
+                        .filter(|cap_id| {
+                            !(self.values.contains_key(cap_id)
+                                || self.unrepresented_ids.contains(cap_id))
+                        })
+                        .collect();
+
+                    let env_types: Vec<_> = captured_ids
+                        .iter()
+                        .map(|_| candy_type_ptr.as_basic_type_enum())
+                        .collect();
+
+                    let env_struct_type = self.context.struct_type(&env_types, false);
+
+                    let env_ptr = self.builder.build_malloc(env_struct_type, "").unwrap();
+
+                    for (idx, cap_id) in captured_ids.iter().enumerate() {
+                        let mut value = self.locals.get(cap_id).map(|v| **v);
+
+                        if value.is_none() {
+                            value.replace(
+                                (self
+                                    .values
+                                    .get(cap_id)
+                                    .unwrap_or_else(|| panic!("{cap_id} is not a global")))
+                                .as_basic_value_enum(),
+                            );
+                        }
+
+                        let member = self
+                            .builder
+                            .build_struct_gep(env_struct_type, env_ptr, idx as u32, "")
+                            .unwrap();
+                        self.builder.build_store(member, value.unwrap());
+                    }
+
+                    let mut params: Vec<_> =
+                        parameters.iter().map(|_| candy_type_ptr.into()).collect();
+                    if !captured_ids.is_empty() {
+                        params.push(candy_type_ptr.into());
+                    }
```

Comment:
```suggestion
                    let parameter_count = parameters.len() + usize::from(!captured_ids.is_empty());
                    let params = vec![candy_type_ptr.into(); parameter_count];
```

Plus another one for the responsible HIR ID

Reply:
Will be addressed when working on #629